---
title: "Data processing with scp"
subtitle: ""
author: "Laurent Gatto and Christophe Vanderaa"
date: "2021/08/10"
output:
  xaringan::moon_reader:
    css: [xaringan-themer.css, custom.css]
    nature:
      highlightStyle: github
      highlightLines: true
      ratio: 16:9
      countIncrementalSlides: true
---

```{r theme, echo=FALSE}
xaringanthemer::style_mono_light(base_color = "#23395b")
```

```{r xaringanExtra, echo=FALSE}
xaringanExtra::use_webcam()     ## use w to turn on webcam
xaringanExtra::use_tile_view()  ## use o to get an overview
xaringanExtra::use_fit_screen() ## use Alt-F to fit to screen
xaringanExtra::use_extra_styles(
    hover_code_line = TRUE
)
xaringanExtra::use_panelset()
```

```{r libraries, include = FALSE}
library(scp)
library(tidyverse)
```

```{r knitr_opts, include = FALSE}
knitr::opts_chunk$set(dpi = 300, 
                      fig.width = 7,
                      message = FALSE, 
                      warning = FALSE)
```

class: middle
name: cc-by

### Get the slides at [https://bit.ly/scp_processing](https://bit.ly/scp_processing)

These slides are available under a **creative common
[CC-BY license](http://creativecommons.org/licenses/by/4.0/)**. You are
free to share (copy and redistribute the material in any medium or
format) and adapt (remix, transform, and build upon the material) for
any purpose, even commercially
<img height="20px" alt="CC-BY" src="https://raw.githubusercontent.com/UCLouvain-CBIO/scp-teaching/main/img/cc1.jpg" />.

???
In this presentation, I will discuss about data processing and will 
show you how to perform two important steps: quality control at 
feature level and quality control at sample level. I will also briefly
mention and demonstrate how to perfor dimension reduction as an example
of downstream analysis.

The slides are available at the given link and are shared under CC-BY 
license. 

---

class: middle, inverse, center

# How to process single-cell proteomics data? 

???
A burning question you might have once you acquired you single-cell
proteomics data is: how should I process single-cell proteomics data?

---

class: middle

## How to process single-cell proteomics data? 

Overview of the workflow suggested in the SCoPE2 seminal paper [1].

```{r echo=FALSE, out.width='50%', fig.align = 'center'}
knitr::include_graphics('./figs/scp_processing_workflow.svg')
```

A full reproduction of the workflow using `scp` and `QFeatures` is 
available from our 
[preprint](http://dx.doi.org/10.1101/2021.04.12.439408) [2] and 
[replication vignette](https://uclouvain-cbio.github.io/SCP.replication/articles/SCoPE2.html).

<p style="color:grey;font-size:0.75em;">
[1] Specht, Harrison, Edward Emmott, Aleksandra A. Petelski, R. Gray 
Huffman, David H. Perlman, Marco Serra, Peter Kharchenko, Antonius 
Koller, and Nikolai Slavov. 2021. “Single-Cell Proteomic and 
Transcriptomic Analysis of Macrophage Heterogeneity Using SCoPE2.” 
Genome Biology 22 (1): 50.
<br>
[2] Vanderaa, Christophe, and Laurent Gatto. 2021. “Utilizing Scp for
the Analysis and Replication of Single-Cell Proteomics Data.” bioRxiv.
</p> 

???
Answering this question is not trivial. While further research is 
required to provide principled guidelines about single-cell proteomics 
data processing, we can already rely on existing processing workflows.
For instance, I show here the workflow from the SCoPE2 seminal paper 
published by Nikolai's lab. 

There are many different steps. Steps shown in yellow are steps commonly
applied to bulk proteomics and are therefore available from the `QFeatures`
package. However, some steps are specific to single cells and the functions
to performs them are available in the `scp` package. 

Due to time constrains, I won't cover all steps, but you can find a full
reproduction of this workflow in our replication vignette and a discussion
in our preprint. 

---

class: middle, center, inverse

# How to perform PSM quality control?

???
Let's start with one of the steps and let me show you how to perform 
PSM quality control

---

class:

## PSM quality control

.panelset[
.panel[.panel-name[Description]

<br> PSM quality control removes low-quality PSMs to **improve the
reliability** of downstream analysis results.

<br> 
**Example**: sample to carrier ratio

The sample to carrier ratio is the signal from samples divided by the
signal of the carrier (tens to hundreds of cell equivalent).

PSMs with high ratios indicate issues during acquisition and/or 
quantification and need to be removed. 

]
.panel[.panel-name[Plot]

Distribution of the sample to carrier ratio computed for all PSMs.        

```{r, out.width = "80%", fig.asp = 0.4, echo = FALSE, fig.align = 'center'}
data("scp1")
scope2Data <- scp1[, , 1:3]
scope2Data <- computeSCR(scope2Data, i = 1:3, colvar = "SampleType",
                         samplePattern = "Macrophage|Monocyte|Blank",
                         carrierPattern = "Carrier", rowDataName = "SCR")
data.frame(rbindRowData(scope2Data, i = 1:3)) %>% 
    ggplot() +
    aes(x = SCR) +
    geom_histogram() +
    geom_vline(xintercept = 0.1, lty = "dashed") +
    scale_x_log10() +
    xlab("Signal to carrier ratio")
```


]
.panel[.panel-name[Code]

Code for computing signal to carrier ratio for each PSM:

```{r, eval = FALSE}
scope2Data <- computeSCR(scope2Data, i = 1:3, colvar = "SampleType",
                         samplePattern = "Macrophage|Monocyte|Blank",
                         carrierPattern = "Carrier", rowDataName = "SCR")
```

Minimal code for plotting:

```{r, eval = FALSE}
rd <- rbindRowData(scope2Data, i = 1:3)
ggplot(data.frame(rd)) +
    aes(x = SCR) +
    geom_histogram() +
    scale_x_log10()
```

Code for filtering:

```{r, eval = FALSE}
scope2Data <- filterFeatures(scope2Data, ~ SCR < 0.1)
```

]
]

???
### Description 

The objective of PSM quality control is to identify and remove 
low-quality PSMs in order to improve the reliability of the downstream
analysis results.

Let me use an example of quality control: the sample to carrier
ratio which is the signal from samples divided by the signal from
carriers. Carriers can contain tens to hundreds of cell equivalent. The logic 
is: when a PSM that exhibits a high ratio, meaning the signal in samples
becomes close to the signal in carriers or even exceeds it, then this 
means that an issues has occurred during acquisition and/or 
quantification and we need to remove those PSMs. 

### Plot

Let's see an example distribution of the sample to carrier ratios.
Most ratios are close to 1% as expected by the experimental design, 
but you may have noticed there is a trail with much higher ratios. 
Those indicate an issue and will be removed.

### Code

Let's quikcly see how to do that. More details about the code are 
provided in the replication vignette. First, we need to compute the the sample to carrier ratio 
using `computeSCR()`. It takes a `QFeatures` object and we tell how the
samples and how the carriers are defined. The results are stored as 
feature annotation, here called `SCR`. 

Then, the second code shunk plots the distribution. We extract the feature 
annotations and provide it to the `ggplot()` function for plotting. 
This will create the plot I just showed. 

The last command is how you can apply the filter. The `filterFeatures()`
function takes the `QFeatures` object and a filtering statement, in this
case we keep only features that have a ratio lower than 10\%.

---

class: middle, center, inverse

# How to perform single-cell quality control?

???
Similarly to what we just saw, I will now demonstrate the quality 
control on single-cell samples.

---

class:

## Single-cell quality control

.panelset[
.panel[.panel-name[Description]

<br> Single-cell quality control removes failed cell to **improve the
reliability** of downstream analysis results.

<br> 
**Example**: median coefficient of variation

The coefficient of variation measures the robustness of quantification
for a protein in a sample. Taking the median across a single cell 
provides an estimate of the robustness of quantification within that
single cell.

Single cells with a high median coefficient of variation indicate 
issues during acquisition and need to be removed. 

]
.panel[.panel-name[Plot]

Distribution of the median coefficient of variation computed for all 
single-cell and blank samples. 

```{r, out.width = "70%", fig.asp = 0.5, echo = FALSE, fig.align = 'center'}
data("scp1")
scope2Data <- scp1
scope2Data <- subsetByColData(scope2Data, scope2Data$SampleType %in% c("Macrophage", "Monocyte", "Blank"))
scope2Data <- medianCVperCell(scope2Data, i = 1:3, 
                              groupBy = "Leading.razor.protein", 
                              colDataName = "medianCV",
                              nobs = 6, 
                              norm = "SCoPE2")
data.frame(colData(scope2Data)) %>% 
    ggplot() +
    aes(x = medianCV,
        fill = SampleType) +
    geom_histogram() +
    scale_x_log10() +
    geom_vline(xintercept = 0.5, lty = "dashed")
```

]
.panel[.panel-name[Code]

Code for computing the median coefficient of variation:

```{r, eval = FALSE}
scope2Data <- medianCVperCell(scope2Data, i = 1:3, groupBy = "Leading.razor.protein", 
                              colDataName = "medianCV", nobs = 6,  norm = "SCoPE2")
```

Code for plotting:

```{r, eval = FALSE}
cd <- colData(scope2Data)
ggplot(data.frame(cd)) +
    aes(x = medianCV, fill = SampleType) +
    geom_histogram() +
    scale_x_log10() +
    geom_vline(xintercept = 0.5, lty = "dashed")
```

Code for filtering:

```{r, eval = FALSE}
scope2Data <- subsetByColData(scope2Data, scope2Data$medianCV < 0.5)
```

]
]

???
### Description 

The objective of single-cell quality control is to identify and remove 
low-quality cells in order to improve the reliability of the downstream
analysis results, just like for feature quality control.

Again, let me use an example: The coefficient of variation measures 
the robustness of quantification for a protein in a sample. Taking the
median across a single cell provides an estimate of the robustness of 
quantification within that single cell. Single cells with a high 
median coefficient of variation indicate issues during acquisition and
need to be removed.

### Plot

We can plot the distribution of the median coefficient of variation 
computed for all single-cells. Including blank samples, that should
theoretically contain mostly noise signal, helps to define a threshold
where single cells are considered too noisy. Here you can see we set 
the threshold at 0.5 to exclude all blank samples but this excludes 
also a few low-quality single cells. 

### Code 

I here show you the code to perform the quality control, again split
in three chunks. The first chunk computes the median coefficient of
variation using the `medianCVperCell` function on a `QFeatures` object.
Various arguments allow to control how the coefficients are computed. 
The computed coefficients are directly stored in the `QFeatures` 
object as sample annotation. 

The second chunk shows how to create the plot I just showed. We retrieve
the sample annotation and provide it to the `ggplot` function to 
create the histogram. 

The last chunk applies the filtering. `subsetByColData()` means that
we take a subset of the samples based on the sample annotation. Here, we 
select all samples that have a median coefficient of variation lower 
than 0.5, the threshold we defined based on the blanks. 

And so this is how to perform two of the many steps of the workflow. Again, 
we a comprehensive demonstration is provided in the replication vignette.

---

class: middle, inverse, center

# What's next after data processing?

???
You may ask yourself: what can I do after data processing? 

---

class: 

## What's next after data processing?

.panelset[
.panel[.panel-name[Description]

Downstream analysis is were the fun begins!

Common analyses: dimension reduction, cluster analysis, cluster 
annotation, differential protein abundance analysis, ...

Many methods are readily applicable to the data thanks to the 
**Bioconductor** 
[`SingleCellExperiment`](https://www.bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html)
container [1]

**Example**: dimension reduction using the `scater` R package. 

<br>
<p style="color:grey;font-size:0.75em;">
[1] Amezquita, Robert A., Aaron T. L. Lun, Etienne Becht, Vince J. Carey, Lindsay N. Carpp, Ludwig Geistlinger, Federico Martini, et al. 2019. “Orchestrating Single-Cell Analysis with Bioconductor.” Nature Methods, December, 1–9.
</p> 

]
.panel[.panel-name[Plot]

t-SNE plot of the protein data. 

```{r, out.width = "80%", fig.asp = 0.4, echo = FALSE, fig.align = 'center'}
data("scp1")
scope2Data <- scp1
library(scater)
prots <- getWithColData(scope2Data, "proteins")
prots <- runTSNE(impute(prots, method = "zero"), 
                 exprs_values = 1,
                 perplexity = 5, 
                 name = "TSNE")
plotTSNE(prots, colour_by = "Set")
```

]
.panel[.panel-name[Code]

Code for getting the protein data with sample annotations:

```{r, eval = FALSE}
prots <- getWithColData(scope2Data, "proteins")
```

Code for computing the tSNE (after imputation by zero):

```{r, eval = FALSE}
prots <- impute(prots, method = "zero")
prots <- runTSNE(prots, exprs_values = 1, perplexity = 5,  name = "TSNE")
```

Code for plotting:

```{r, eval = FALSE}
plotTSNE(prots, colour_by = "Set")
```

]
]

???
### Description

Once your data is processed, you are ready to perform downstream 
analyses and that's were the fun begins! Common downstream analyses are
for instance dimension reduction, cluster analysis, cluster annotation,
differential protein abundance analysis and many more. 

Most of the methods to perform single-cell analyses are actually 
already available. We can directly apply those methods
to our data because our data is contained in `SingleCellExperiment`
objects within the `QFeatures` object. 

As an example, let me demonstrate how to apply dimension reduction  
using the `scater` package. 

### Plot

Here is a tSNE plot generated from an example protein data obtained by
aggregating the PSMs to proteins without further processing. You can 
see on this plot that the main source of variability is the acquisition
batch, and this is often not desirable. So, a thorough data processing is critical for principled data 
analysis in order to generate robust biological knowledge. If we do not correctly 
model the data, we may pass through interesting information or focus 
on confounding effects. 

### Code

Again, here is a bit of code to perform dimension reduction. First, we extract
the data of interest from the `QFeatures` object. In this case, we 
want the protein data using the `getWithColData(à` function that will
also transfer all the available sample annotations that we use for 
colouring the plot. 

Next, we compute the tSNE after a little imputation of missing data by
zero values. `runTSNE()` is a function from the `scater` package. 

Finally, we generate the tSNE plot and colour it by the acquisition 
set. Again, `plotTSNE()` is a function from `scater`. 

I hope this presentation could help you understand how to perform 
data processing using our software, namely quality control, and what 
you could do with your data after processing.

---

class: middle, inverse, center

# Exercise

???
Let's know test your understanding with a small exercise.

---

class: middle

#### Given the distribution of the median coefficient of variation per cell, what command would you use to remove low-quality cells?

```{r echo=FALSE, out.width='50%', fig.align = 'center'}
knitr::include_graphics('./figs/scp_processing_exercise.svg')
```

```{r, eval = FALSE}
1. subsetByColData(scope2Data, scope2Data$medianCV < 0.25)
2. subsetByColData(scope2Data, scope2Data$medianCV > 0.25)
3. subsetByColData(scope2Data, scope2Data$medianCV < 0.32)
4. subsetByColData(scope2Data, scope2Data$medianCV > 0.32)
5. subsetByColData(scope2Data, scope2Data$medianCV < 0.45)
6. subsetByColData(scope2Data, scope2Data$medianCV > 0.45)
```

???
Given the distribution of the median coefficient of variation per cell,
what command would you use to remove low-quality cells? 

Command 1, 2, 3, 4, 5 or 6? You can pause the video to think about it... 
The solution is command 3. Setting a threshold at 0.32 best separates
between blank samples and the single-cells and we are interested to
keep samples with a low median coefficient of variation. 

---

class: middle

### Further information

* Check out our vignette where we fully replicated the SCoPE2 workflow:
  https://uclouvain-cbio.github.io/SCP.replication/articles/SCoPE2.html
* Check out our prepint where we discuss the current challenges in 
  single-cell proteomics data anylysis: 
  http://dx.doi.org/10.1101/2021.04.12.439408)
* Try it yourself in this online workshop: 
  https://lgatto.github.io/QFeaturesScpWorkshop2021/

### Funding

Fonds de la Recherche Scientifique (FNRS), Belgium
    
???
I hope you found the correct solution. If you're looking for more 
detailed information, you can have a look at our replication vignette
and our preprint. If your interested to get your hands on our software,
check out our online workshop! Thank you very much for watching and 
see you in another video.

