<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>What is scp?</title>
    <meta charset="utf-8" />
    <meta name="author" content="Laurent Gatto and Christophe Vanderaa" />
    <script src="what_is_scp_files/header-attrs-2.10/header-attrs.js"></script>
    <script src="what_is_scp_files/xaringanExtra-webcam-0.0.1/webcam.js"></script>
    <script id="xaringanExtra-webcam-options" type="application/json">{"width":"200","height":"200","margin":"1em"}</script>
    <link href="what_is_scp_files/tile-view-0.2.6/tile-view.css" rel="stylesheet" />
    <script src="what_is_scp_files/tile-view-0.2.6/tile-view.js"></script>
    <script src="what_is_scp_files/xaringanExtra_fit-screen-0.2.6/fit-screen.js"></script>
    <link href="what_is_scp_files/xaringanExtra-extra-styles-0.2.6/xaringanExtra-extra-styles.css" rel="stylesheet" />
    <link rel="stylesheet" href="xaringan-themer.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# What is scp?
### Laurent Gatto and Christophe Vanderaa
### 2021/08/10

---







class: middle
name: cc-by

### Get the slides at [http://bit.ly/what_is_scp](http://bit.ly/what_is_scp)

These slides are available under a **creative common
[CC-BY license](http://creativecommons.org/licenses/by/4.0/)**. You are
free to share (copy and redistribute the material in any medium or
format) and adapt (remix, transform, and build upon the material) for
any purpose, even commercially
&lt;img height="20px" alt="CC-BY" src="https://raw.githubusercontent.com/UCLouvain-CBIO/scp-teaching/main/img/cc1.jpg" /&gt;.


---
class: middle, center, inverse

# What is scp?

---

class: middle

### What is scp?

- **scp** stands for *single cell proteomics*.

- **scp** is a piece of software to process, analyse and interpret
  mass spectrometry-based *single cell protemics* data.

- It is an [R](http://r-project.org) package - it is implemented in
  the R language for statistical computing and is operated through the
  R environment.

- It is part of
  [Bioconductor](http://www.bioconductor.org/packages/scp) - it is
  interoperable and thus benefits from numerous software for the
  analysis of omics data, such as proteomics or single cell
  transcriptomics, ...

---
class: middle, center, inverse

# How to get started with scp?

---
class: middle

### How to get started with scp?

The **scp**

- Pre-print ([Vanderaa and Gatto, bioRXiv,
  2021](https://www.biorxiv.org/content/10.1101/2021.04.12.439408v1))
- Webpage and vignettes at https://UCLouvain-CBIO.github.io/scp
- On-line tutorial at https://lgatto.github.io/QFeaturesScpWorkshop2021/

---
class: middle, center, inverse

# Is it like ... (insert your favourite software)?

---
class: middle

### Is it like ... (insert your favourite software)?

No, **scp** will give you more flexibility than any other software and
functionality dedicated to mass spectrometry-based single cell
proteomics.

---
class: middle, center, inverse

# Are there any alternatives to scp?

---
class: middle

### Are there any alternatives to scp?

The closest to **scp** is the python-based `SCeptre` ([Schoof *et
al.*, 2021](https://www.nature.com/articles/s41467-021-23667-y)),
although it isn't as advances as **scp**.

---
class: middle, center, inverse

# Do we need dedicated single cell data analysis/software?

---
class: middle

### Do we need dedicated single cell data analysis/software?

- batch effects
- missing values
- and their interaction
- ...

---
class: middle, center, inverse

# How to install scp (and its dependencies)?

---
class: middle

### How to install scp (and its dependencies)?

Install R and RStudio and execute


```r
if (!require("BiocManager"))
    install.packages("BiocManacer")

BiocManager::install("scpdata")
BiocManager::install("scp")
```

---
class: middle, center, inverse

# Can I test it without installing it?

---
class: middle

### Can I test it without installing it?

Yes! See https://lgatto.github.io/QFeaturesScpWorkshop2021/

---
class: middle, center, inverse

# Do I need to know R?

---
class: middle

### Do I need to know R?

Yes!

---
class: middle, center, inverse

# Do I need to be a programmer/statistician/bioinformatician ...?

---
class: middle

### Do I need to be a programmer/statistician/bioinformatician ...?

No, you **only** need to be keen to analyse and understand you data!

---
class: middle, center, inverse

# Where can I get help?

---
class: middle

### Where can I get help?

- The Bioconductor support site: http://support.bioconductor.org
- The **scp** github issues: http://github.com/UCLouvain-CBIO/scp/issues
- Regular workshops we teach

---
class: middle, center, inverse

# Next steps

---
class: middle

### Next steps

- Try it out, locally or in the cloud.
- Check the documentation.
- Ask questions!

### Funding

Fonds de la Recherche Scientifique (FNRS), Belgium


    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"ratio": "16:9",
"countIncrementalSlides": true
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
